File: api/java-client/src/main/java/screencastnarrator/Highlight.java
package screencastnarrator;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;

public class Highlight {

    private Highlight() {}

    public static void drawHighlight(Page page, Locator locator, SharedConfig config, CdpVideoRecorder recorder) {
        locator.evaluate(config.highlight().getScrollJs());
        page.evaluate(config.highlight().getScrollWaitJs());
        locator.evaluate(config.resolvedDrawJs());
        waitWithAcks(page, config.highlight().getAnimationSpeedMs() + config.highlight().getDrawWaitMs(), recorder);
    }

    public static void removeHighlight(Page page, SharedConfig config, CdpVideoRecorder recorder) {
        page.evaluate(config.highlight().getRemoveJs());
        waitWithAcks(page, config.highlight().getRemoveWaitMs(), recorder);
    }

    public static void highlight(Page page, Locator locator, SharedConfig config, CdpVideoRecorder recorder) {
        drawHighlight(page, locator, config, recorder);
        removeHighlight(page, config, recorder);
    }

    public static void highlight(Page page, Locator locator, SharedConfig config) {
        highlight(page, locator, config, null);
    }

    private static void waitWithAcks(Page page, int totalMs, CdpVideoRecorder recorder) {
        if (recorder == null || totalMs <= 0) {
            page.waitForTimeout(totalMs);
            return;
        }
        int elapsed = 0;
        int interval = 40;
        while (elapsed < totalMs) {
            int wait = Math.min(interval, totalMs - elapsed);
            page.waitForTimeout(wait);
            elapsed += wait;
            recorder.ackLatestFrame();
        }
    }
}

File: api/python-client/src/screencast_narrator_client/highlight.py
"""Element highlighting: draws an animated elliptical circle around a Playwright locator."""

from __future__ import annotations

from screencast_narrator_client.shared_config import SharedConfig


def draw_highlight(page, locator, config: SharedConfig) -> None:
    hl = config.highlight
    locator.evaluate(config.resolved_scroll_js)
    page.evaluate(config.resolved_scroll_wait_js)
    locator.evaluate(config.resolved_draw_js)
    page.wait_for_timeout(hl.animation_speed_ms + hl.draw_wait_ms)


def remove_highlight(page, config: SharedConfig) -> None:
    page.evaluate(config.resolved_remove_js)
    page.wait_for_timeout(config.highlight.remove_wait_ms)


def highlight(page, locator, config: SharedConfig) -> None:
    draw_highlight(page, locator, config)
    remove_highlight(page, config)
